File: src/main/java/be/breina/openrgb/command/CommandResponse.java
package be.breina.openrgb.command;

import static java.nio.ByteOrder.*;

import java.nio.ByteBuffer;
import java.util.Arrays;

public class CommandResponse {

    private final int deviceId;
    private final CommandId commandId;
    private final int size;

    public static CommandResponse from(byte[] command) {
        return new CommandResponse(command);
    }

    private CommandResponse(byte[] command) {
        ByteBuffer byteBuffer = ByteBuffer.wrap(command).order(LITTLE_ENDIAN);

        if (Arrays.mismatch(CommandConstants.COMMAND_HEADER, byteBuffer.array()) != CommandConstants.COMMAND_HEADER.length) {
            throw new IllegalStateException("Wrong length!");
        }

        deviceId = byteBuffer.getInt(4);
        commandId = CommandId.from(byteBuffer.getInt(8));
        size = byteBuffer.getInt(12);
    }

    public int getDeviceId() {
        return deviceId;
    }

    public CommandId getCommandId() {
        return commandId;
    }

    public int getSize() {
        return size;
    }
}

File: src/main/java/be/breina/openrgb/command/CommandResponse.kt
package be.breina.openrgb.command

import be.breina.openrgb.command.CommandId
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*

class CommandResponse private constructor(command: ByteArray) {
    val deviceId: Int
    val commandId: CommandId
    val size: Int

    companion object {
        fun from(command: ByteArray): CommandResponse = CommandResponse(command)
    }

    init {
        val byteBuffer = ByteBuffer.wrap(command).order(ByteOrder.LITTLE_ENDIAN)
        check(Arrays.mismatch(CommandConstants.COMMAND_HEADER, byteBuffer.array()) == CommandConstants.COMMAND_HEADER.size) { "Wrong length!" }
        deviceId = byteBuffer.getInt(4)
        commandId = CommandId.Companion.from(byteBuffer.getInt(8))
        size = byteBuffer.getInt(12)
    }
}